src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -211,12 +215,51 @@ protected override void OnSysCall(uint method)
                 Push(Convert(returnValue));
         }
 
+        internal void Trace()
+        {
+            if (traceDebugSink != null)
+            {
+                var storages = InvocationStack
+                    .Select(ec => ec.GetScriptHash())
+                    .Distinct()
+                    .SelectMany(GetStorages);
+
+                traceDebugSink.Trace(State, InvocationStack, storages);
+            }
+
+            IEnumerable<(UInt160 scriptHash, byte[] key, StorageItem item)> GetStorages(UInt160 scriptHash)
```

Comment:
Outside the method?

Reply:
Since GetStorages is only used by trace, I prefer making it a local function. Do we have local style preferences against local functions in Neo project?

Reply:
I think that we haven't got any local function xD

Reply:
neo-express and debugger have them. I find them very useful in cases like this where you can't express a LINQ clause in an anonymous lambda expression. 

If it's a big deal, I'll fold the local function into an anonymous lambda block for SelectMany. I'd rather not introduce a helper method that's only called in one place. Let me know if you want me to make that change